Bolt 3.3+ Amazon S3 Remote Filesystem

GawainLynch opened the pull request:
These are the starter docs for using Amazon S3 as a remote filesystem in Bolt :tada: 

I would like to point out that a huge amount of work on the Bolt & Filesystem repositories, and most of the example code here is thanks to @CarsonF … seriously, if you knew the hours that have gone into making this happen! :smile: :+1:

CarsonF commented on docs/performance/remote-filesystem/introduction.md:
```diff
@@ -0,0 +1,14 @@
+geographically relevant data centres around the world.
+
+To support this seamlessly in the Bolt application, our filesystem service
+provides support for Remote Filesystem adapters individicual filesystems.
```
typo.

GawainLynch replied:
Done

CarsonF commented on docs/performance/remote-filesystem/amazon-s3.md:
````diff
@@ -0,0 +1,396 @@
+
+```bash
+$ composer require league/flysystem-aws-s3-v3:^1.0 league/flysystem-cached-adapter:^1.0 league/flysystem-memory:^1.0
+```
````
`flysystem-memory` isn't required.

GawainLynch replied:
Done

CarsonF commented on docs/performance/remote-filesystem/amazon-s3.md:
```diff
@@ -0,0 +1,396 @@
+### Caching Libraries
+----------------------
+
+While other cache targets are optional, right now this example will be base on Redis.
```
Maybe say that any PSR-6 or Doctrine Cache instances can be used?

Also there's a reason I used Redis. The cache needs to be shared amongst all the nodes. We are using Elasticache for this.

GawainLynch replied:
Done

CarsonF commented on docs/performance/remote-filesystem/amazon-s3.md:
```diff
@@ -0,0 +1,396 @@
+<code>psr-4</code> section of your <code>composer.json</code> file, remember to
+re-run <code>composer dump-autoload</code> to update the cached autoloading
+files.
+</p>
```
This is "advanced" but we are still going over autoloading?...

GawainLynch replied:
I was, and still am, on the fence on this … and increasingly leaning towards _leaving_ it. It doesn't provide _"documentation"_ that is repeated elsewhere, just a _reminder_ for those that are prone to forget these things, i.e. they don't think like Carson. :wink:

bobdenotter replied:
I'd say it should stay in. People are very good at just skipping over stuff they don't need, so there's no real harm in keeping it. Alternatively, for a bunch of people this is not trivial. :-)

CarsonF commented on docs/performance/remote-filesystem/amazon-s3.md:
```diff
@@ -0,0 +1,396 @@
+
+                return new Sdk($options);
+            }
+        );
```
Let's skip the `Sdk` part. S3 clients can be created directly - `new S3Client`

GawainLynch replied:
Done

CarsonF commented on docs/performance/remote-filesystem/amazon-s3.md:
```diff
@@ -0,0 +1,396 @@
+#### Amazon Credentials
+
+In this example we use `CredentialProvider::defaultProvider()` to get your
+Amazon credentials.
```
I can't think of a case where the defaults wouldn't be wanted since it basically checks everywhere. And if you don't follow their defaults then other tools have to have the creds passed in manually too.

Let's just leave the code out. I think it's still good to list where the defaults check.

CarsonF commented on docs/performance/remote-filesystem/amazon-s3.md:
```diff
@@ -0,0 +1,396 @@
+Amazon credentials.
+
+This credential provider first checks in order:
+  * Amazon specific environment variables
```
`AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`

GawainLynch replied:
Done

CarsonF commented on docs/performance/remote-filesystem/amazon-s3.md:
```diff
@@ -0,0 +1,396 @@
+
+This credential provider first checks in order:
+  * Amazon specific environment variables
+  * "default" profile in ~/.aws/credentials
```
Or the profile specified by `AWS_PROFILE` env var.

GawainLynch replied:
Done

CarsonF commented on docs/performance/remote-filesystem/amazon-s3.md:
```diff
@@ -0,0 +1,396 @@
+  * "profile default" profile in ~/.aws/config (which is the default profile of
+    AWS CLI)
+  * Tries to make a GET Request to fetch credentials if the Ecs environment
+    variable is presented
```
Simplify to `ECS environment variables`

GawainLynch replied:
Done

CarsonF commented on docs/performance/remote-filesystem/amazon-s3.md:
```diff
@@ -0,0 +1,396 @@
+    AWS CLI)
+  * Tries to make a GET Request to fetch credentials if the Ecs environment
+    variable is presented
+  * Checks for EC2 instance profile credentials
```
Change to "Fetches credentials for the EC2 instance's IAM Role".

Would also note that "If you are using EC2 instances then this is the recommended way to provide credentials".

GawainLynch replied:
Done

CarsonF commented on docs/performance/remote-filesystem/amazon-s3.md:
```diff
@@ -0,0 +1,396 @@
+                'credentials' => $credentials,
+            ];
+
+            return $app['aws']->createS3($options);
```
to reiterate
```php
return new S3Client([
  'region' => 'us-east-1', // bucket's region
  'version' => '2006-03-01',
]);
```

GawainLynch replied:
Done

CarsonF commented on docs/performance/remote-filesystem/amazon-s3.md:
```diff
@@ -0,0 +1,396 @@
+<br>
+Saving of thngs like secrect keys to code that gets committed to version
+control is bad security practice.
+</p>
```
Nah, drop this section. K.I.S.S.

GawainLynch replied:
Done

CarsonF commented on docs/performance/remote-filesystem/amazon-s3.md:
```diff
@@ -0,0 +1,396 @@
+
+The second parameter is the version strategy applicable to the package, the
+value in the final parentheses **must** be the name of the package you're
+replacing.
```
It must be the name of the filesystem you are replacing. But it should match the package name as well, or the fs asset plugin will break.

GawainLynch replied:
Done

CarsonF commented on docs/performance/remote-filesystem/amazon-s3.md:
```diff
@@ -0,0 +1,396 @@
+replacing.
+
+In this example, 'files' is package name, that is package that is responsible
+for the contents of `{web root}/files/`.
```
This is confusing, as it's not a local directory anymore.

GawainLynch replied:
Clarified, and removed `{web root}` also

CarsonF commented on docs/performance/remote-filesystem/amazon-s3.md:
```diff
@@ -0,0 +1,396 @@
+</p>
+
+
+### Assets
```
I would move this below the filesystem section. Filesystem is the next logical jump from AWS configuration.

GawainLynch replied:
Done

CarsonF commented on docs/performance/remote-filesystem/amazon-s3.md:
```diff
@@ -0,0 +1,396 @@
+                'asset.packages',
+                function ($packages, $app) {
+                    $package = new UrlPackage(
+                        'http://example.s3-website.eu-central-1.amazonaws.com',
```
CloudFront CDN url can also be used here.

CarsonF commented on docs/performance/remote-filesystem/amazon-s3.md:
```diff
@@ -0,0 +1,396 @@
+
+Finally, you must extend `$app['filesystem']`, and call `mountFilesystems()`
+using an assicative array of the filesystem(s) you want to override using the
+`$app['filesystem.s3']` protected callable.
```
I would flip this order, so the user can follow why. "We want to override the "files" filesystem." -> We need an S3 adapter for that -> The adapter should be cached for performance.

Also you are mixing general documentation with specifics from the example below. Maybe it would be better to have those separated. Or maybe have more comments in the example below?

GawainLynch replied:
Rehashed it, but I'll get you to have a re-read when I push the `asdf` commit

CarsonF commented:
Here's the IAM policy that you'll need to attach to the IAM role / user.
```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "s3:HeadBucket",
                "s3:GetBucketAcl"
                "s3:HeadObject",
                "s3:GetObject",
                "s3:GetObjectAcl",
                "s3:GetObjectVersion",
                "s3:GetObjectVersionAcl",
                "s3:PutObject",
                "s3:PutObjectAcl",
                "s3:PutObjectVersionAcl",
                "s3:DeleteObject",
                "s3:DeleteObjectVersion"
            ],
            "Resource": "arn:aws:s3:::bucket-name/*"
        }
    ]
}
```

GawainLynch commented:
@CarsonF I have pushed a commit "Review updates" that I'll squash pre-merge … so let me have it :wink:

CarsonF commented on docs/performance/remote-filesystem/amazon-s3.md:
```diff
@@ -0,0 +1,377 @@
+#### Filesystem Cache Factory
+
+Create a protected callable `$app['filesystem.cache_factory']` that when
+invoked returns a class extending `\League\Flysystem\Cached\CachedAdapter`.
```
This needs to be our subclass instead of Leagues actually

GawainLynch replied:
Done

CarsonF commented on docs/performance/remote-filesystem/amazon-s3.md:
```diff
@@ -0,0 +1,377 @@
+
+The second parameter is the version strategy applicable to the package, the
+value in the final parentheses **must** be the name of the package you're
+replacing, and **must** match the package name used in `addPackage()` as well.
```
> **must** be the name of the package you're replacing

**must** be the mount point of the filesystem you are changing.

GawainLynch replied:
Done

CarsonF commented on docs/performance/remote-filesystem/amazon-s3.md:
````diff
@@ -0,0 +1,377 @@
+
+```bash
+$ composer require predis/predis:^1.1
+```
````
The cache needs to be shared between servers, else bad things happen. We recommend using Elasticache for a shared Redis instance, and using the Predis library to connect to it.

GawainLynch replied:
Done, with links :+1:

GawainLynch commented:
> Still missing the IAM example too, unless you think it's not needed.

Yeah, I wanted to put it in, I didn't know what to do with it! 

I have no objections, just haven't had the time to combine reading up Amazon API docs as well :rofl:

GawainLynch commented:
OK, heavily updated mate.

bobdenotter commented on docs/performance/remote-filesystem/introduction.md:
```diff
@@ -0,0 +1,14 @@
+Introduction
+============
+
+On more complex, and high-volumne sites, using Content Delivery Networks (CDN)
```
Typo: `volumne`

GawainLynch replied:
Fixed

GawainLynch commented:
> Still missing the IAM example too, unless you think it's not needed.

Had a very quick read up, modified your sentence slightly and added the JSON for the policy.

siada commented:
Need to add a note about requiring `symfony/browser-kit` as this doesn't work without it

GawainLynch commented:
> Need to add a note about requiring `symfony/browser-kit` as this doesn't work without it

What requires BrowserKit?

siada commented:
I'm not 100% sure of the root, but the error itself is thrown by `symfony\http-kernel`

GawainLynch commented:
I just checked on my local install, that does **not** have BrowserKit installed, it work, and is only being invoked in the whole tree by tests … something you've configured locally?

siada commented:
Hmm... I thought it might be this as the error only occurs when the AwsServiceProvider is included in .bolt.php

GawainLynch commented:
How recent is the `bolt/bolt` 3.3-beta you're working with?

siada commented:
I'm running beta-10

GawainLynch commented:
OK, so not what I was thinking then. :+1:

CarsonF commented:
@siada did you update your composer requirements (as mentioned in doc)?

siada commented:
Yes, I executed:

CarsonF commented:
Well then did you change AwsServiceProvider to not use Predis then? ;)

siada commented:
So I just tested it on a completely brand new installation, this is the jist of my process:
`composer create-project bolt/composer-install:^3.2`
then update composer.json to be 3.3-beta10 instead of ^3.2
`composer require league/flysystem-aws-s3-v3:^1.0 league/flysystem-cached-adapter:^1.0`
`mkdir -p src/Provider`
`vi src/Provider/AwsServiceProvider.php`
Paste in example from amazon-s3.md
create .bolt.yml and add the AwsServiceProvider
`composer dumpautoload`
try load homepage -> ![image](https://user-images.githubusercontent.com/177663/27299257-a946ec30-5523-11e7-802e-c999861d3d46.png)

so it looks like somethings not right? I fully understand that I've not done any of the configuration for the S3 bucket, but it's not even getting as far as that as there is a package which is considered 'missing'.

CarsonF commented:
@siada you didn't require `predis/predis` but the example `AwsServiceProvider` uses it. https://github.com/bolt/docs/pull/740/files#diff-04e49214ef18381423b26544246c2a63R257

siada commented:
Oh... that did actually get rid of the error 😹 why would installing predis fix an error with http-kernel\Client trying to load browserkit ?

CarsonF commented:
`Predis\Client` is trying to be loaded from a call made within the BrowserKit namespace. It's then trying to give you a list of Clients, one of which you "forgot" to import. However, the Predis\Client isn't in that list because the class doesn't exist because you never installed it.

ehough commented:
Just as a point of reference, I was able to follow these instructions about two weeks ago and got my Bolt installation working with S3. I followed the docs carefully and didn't encounter any major hangups along the way. Using PHP 7.1.5, Bolt 3.3 beta, phpredis, and ElastiCache. Really appreciate your great work on this.

GawainLynch commented:
@siada That is the kind of mistake that I usually make, about twice a day :rofl: … great to know it is sorted, and thank you for testing and giving feedback!

@ehough Thank you for the feedback also!

It is fantastic that the documentation, and our implementation, of the new feature is getting wider testing! :+1: … @CarsonF and I are usually fairly O.C.D. in our approach, but human. 

Thank you all :smile:

siada commented:
No problem! as @ehough pointed out, it does work very well, once I got it running 😁 I think it would be beneficial at some point to get this into the core of Bolt along with support for other providers.

GawainLynch commented:
> I think it would be beneficial at some point to get this into the core of Bolt

Well, the code to enable it to run is in core, and will be released as 3.3 stable very soon. :tada: 

But carrying the dependencies is way out of scope for a base install. This has, since the feature development started, been targeted at people who are building their own Composer project and are comfortable doing so. i.e. if one knows enough to set up an S3 bucket, get it working, write the service providers, and get the loading … one can manage a `composer require` without blinking an eye :wink: 

>  along with support for other providers.

Yes, as time/need/PRs permit we'll add more adaptors to `bolt/filesystem` … but small number of coders versus large amount of project work :smile_cat:

siada commented:
I don't suppose this handles the thumbnail generation does it? I'm trying to keep all images out of my scc repository

CarsonF commented:
@siada yes it does 😄

GawainLynch commented:
**Merge note:** This is waiting on a branch push from @CarsonF … :clock10:

bobdenotter commented:
@CarsonF, @GawainLynch Is this still waiting? I know of a few distinct people who have used this to successfullt implement S3 or even Dropbox Filesystems.

GawainLynch commented:
@bobdenotter: Yeah, still as-per my comment above yours … @CarsonF has a push pending, I gather he just ran out of time before his break. Let's ping him on Slack this afternoon and see if he objects to an as-is merge.

bobdenotter commented:
In the spirit of moving forward, I say we merge this one in as-is. As noted above, it's already quite extensive, and multiple people outside of the core team have reviewed it (by de facto using it to implement this feature in a project). 

BDFL says: It's Good Enough™.
